hybrid/per_func_comp_main.c

```diff
@@ -0,0 +1,257 @@
+    (void) sandbox_end;
+    datacap = cheri_andperm(sandbox_base, CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_STORE |
+                                              CHERI_PERM_LOAD_CAP | CHERI_PERM_STORE_CAP |
+                                              CHERI_PERM_STORE_LOCAL_CAP);
+#else
+    datacap =
+        cheri_ptrperm(sandbox_base, // TODO: define sandbox_base/end
+                      (size_t) sandbox_end - (size_t) sandbox_base,
+                      CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_STORE | CHERI_PERM_LOAD_CAP |
+                          CHERI_PERM_STORE_CAP | CHERI_PERM_STORE_LOCAL_CAP);
+#endif
+    return (datacap);
+}
+
+//- Create a global_sealcap to track the next cheri_otype -//
+void *__capability global_sealcap;
+
+//- Simple type allocator - Takes an otype size; returns the next otype -//
+void *__capability alloc_type(size_t tysz)
+{
+    // TODO: Try with cheri_otype_alloc :: otype_t cheri_otype_alloc(void);
+    // cheri_otype 0~3      : reserved by morello,
+    //             4~ 2^13  : available for userspace
+    // Check how Cheri makes a type: "cheri_otype_t cheri_maketype(sealing_cap, type)"
+    //          ?? -> cheri_maketype return type is otype??
+
+    void *__capability sealcap;
+    sealcap = cheri_maketype(global_sealcap, tysz);
+    __builtin_cheri_cap_type_copy(global_sealcap, sealcap);
+    return sealcap;
+}
+
+/*
+//
+// TODO: test function attributes (ccall, ccallee), with args
+//
+__attribute__((cheri_ccall))
+__attribute__((cheri_method_suffix("_cap")))
+__attribute__((cheri_method_class(main_obj)))
+int sandbox_1_func(int * __capability intcap);
+
+int sandbox_1_func(int * __capability intcap)
+{
+    struct cheri_object sb1_obj;
+
+    sb1_obj.co_codecap = cheriobj_1_codecap;
+    sb1_obj.co_datacap = cheriobj_1_datacap;
+
+    (void)libcheri_invoke(sb1_obj, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+
+    return *intcap + 1;
+}
+
+__attribute__((cheri_ccall))
+__attribute__((cheri_method_suffix("_cap")))
+__attribute__((cheri_method_class(sandbox2_obj)))
+int sandbox_2_func(int * __capability intcap);
+
+int sandbox_2_func(int * __capability intcap)
+{
+    struct cheri_object sb2_obj;
+
+    sb2_obj.co_codecap = cheriobj_2_codecap;
+    sb2_obj.co_datacap = cheriobj_2_datacap;
+
+    (void)libcheri_invoke(sb2_obj, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+
+    return *intcap + 1;
+}
+///
+/// up_to_here
+///
+*/
+
+void sandbox_1_func()
+{
+    printf("--> end_of_sandbox_1_func\n");
+}
+
+void sandbox_2_func()
+{
+    printf("--> end_of_sandbox_2_func\n");
+}
+
+int main() // Remove unused args
+{
+    printf("----------------------------------------\n");
+    printf("CHERI_OTYPE_USER_MIN: %04x\n", CHERI_OTYPE_USER_MIN);
+    printf("CHERI_OTYPE_USER_MAX: %04x\n", CHERI_OTYPE_USER_MAX);
+    printf("----------------------------------------\n\n");
```

Comment:
This looks like debug code? I would remove this if unneeded.

Reply:
To show (potential) user/learner allowed user-defined otype value ranges.

Reply:
I don't think it's worthwhile to have these printed in our example. They look like CHERI macros, and using them in the asserts further down should be sufficient.